### PR TITLE
[posix] handle prefixes in the router advertisement

### PR DESCRIPTION
**Pull request by gjc13**
This change allows users to handle router advertisement from DHCP agent to the Thread network, enabling ISP SLAAC and DHCPv6_PD configuration.

In `wpantund`, we will add a prefix for any external addresses on the interface and ask users to statically config whether to treat the prefix as SLAAC or DHCP. This change can be viewed as an enhancement to the previous approach.

Fixes https://github.com/openthread/ot-br-posix/issues/470.

### CODE REVIEW

**Comment by gjc13** [comment]
@rob-the-dude FYI.

File: tests/scripts/misc/radvd_slaac.conf.in [hunk 1]
```diff
@@ -0,0 +1,12 @@
+interface WPAN { 
+        AdvSendAdvert on;
+        MinRtrAdvInterval 3; 
+        MaxRtrAdvInterval 10;
+        AdvManagedFlag off;
+        prefix 2001:db8:1:2::/64 { 
+                AdvOnLink on; 
+                AdvAutonomous on; 
+                AdvRouterAddr on;
```
Comment (simonlingoogle) [inline tests/scripts/misc/radvd_slaac.conf.in:9]:
```suggestion
                AdvRouterAddr off; 
```

Comment (gjc13) [reply tests/scripts/misc/radvd_slaac.conf.in:9]: We are only listening on RA received on wpan0 interface.

**Comment by gjc13** [comment]
> Are we doing auto prefix configuration on Thread network for all `RA` recevied from backbone? I think it might not be safe and necessary to do so.

I only listen to RA multicasts on the wpan interface. RA on that interface is quite an explicit request to add prefix.

File: src/posix/platform/ra_listener.hpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+     * This method updates the file descriptor sets with file descriptor used by the listener.
+     *
+     * @param[inout]  aReadFdSet   A reference to the read file descriptors.
+     * @param[inout]  aWriteFdSet  A reference to the write file descriptors.
+     * @param[inout]  aErrorFdSet  A reference to the error file descriptors.
+     * @param[inout]  aMaxFd       A reference to the max file descriptor.
+     * @param[inout]  aTimeout     A reference to the timeout.
+     *
+     */
+    void UpdateFdSet(fd_set &aReadFdSet, fd_set &aWriteFdSet, fd_set &aErrorFdSet, int &aMaxFd);
```
Comment (bukepo) [inline src/posix/platform/ra_listener.hpp:85]: use otSysMainloopContext instead?

File: src/posix/platform/ra_listener.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    otLogInfoPlat("RaListener::Init error=%s", otThreadErrorToString(error));
+    return error;
+}
+
+void RaListener::DeInit(void)
+{
+    if (mRaFd != 0)
+    {
+        close(mRaFd);
+        mRaFd = 0;
```
Comment (bukepo) [inline src/posix/platform/ra_listener.cpp:127]: -1?

File: src/posix/platform/ra_listener.cpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    uint8_t  _rsv : 6;
+    uint8_t  mOtherConfiguration : 1;
+    uint8_t  mMananged : 1;
+    uint16_t mRouterLifetime;
+    uint32_t mReachableTime;
+    uint32_t mRetransTimer;
+} OT_TOOL_PACKED_END;
+
+RaListener::RaListener(void)
+    : mRaFd(0)
```
Comment (bukepo) [inline src/posix/platform/ra_listener.cpp:103]:
```suggestion
    : mRaFd(-1)
```

File: src/posix/platform/ra_listener.cpp (hunk 3 above)
Comment (simonlingoogle) [reply src/posix/platform/ra_listener.cpp:127]: I think `-1` is better because `0` is `stdin`.

**Comment by jwhui** [comment]
> This change allows users to forward router advertisement from upstream
> router to the Thread network, enabling ISP SLAAC and DHCPv6_PD
> configuration.

@gjc13 , I think this description is misleading. The Router Advertisement is not coming from an upstream router. Instead, the Router Advertisement is being generated from the same device running OpenThread.

I think what is meant here is that we are using Router Advertisements to learn what prefixes are being assigned to the Thread network and translating them into Thread-specific messages. Is that correct?

**Comment by simonlingoogle** [comment]
> > This change allows users to forward router advertisement from upstream
> > router to the Thread network, enabling ISP SLAAC and DHCPv6_PD
> > configuration.
> 
> @gjc13 , I think this description is misleading. The Router Advertisement is not coming from an upstream router. Instead, the Router Advertisement is being generated from the same device running OpenThread.
> 
> I think what is meant here is that we are using Router Advertisements to learn what prefixes are being assigned to the Thread network and translating them into Thread-specific messages. Is that correct?

This PR makes it possible to setup Thread network prefixes by forwarding certain `Router Advertisements` from Backbone to Thread network?

File: src/posix/platform/netif.cpp [hunk 5]
```diff
@@ -249,6 +253,10 @@ enum
 
 static const size_t kMaxIp6Size = 1536;
 
+#ifdef OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE
+ot::Posix::RaListener sRaListener;
```
Comment (simonlingoogle) [inline src/posix/platform/netif.cpp:257]: Should we have a option for this feature (preferably `ON` by default), so that it's possible to turn it off?

**Comment by jwhui** [comment]
> > I think what is meant here is that we are using Router Advertisements to learn what prefixes are being assigned to the Thread network and translating them into Thread-specific messages. Is that correct?
> 
> This PR makes it possible to setup Thread network prefixes by forwarding certain `Router Advertisements` from Backbone to Thread network?

@simonlingoogle, couple notes:

1. I don't think we are "forwarding" anything.

2. I don't think we are only processing Router Advertisements that are originated by the Thread border router and sent to the Thread interface. We are not processing Router Advertisements originated by any other device or received on any other interface.

What we are doing here is really more of a "hack". We are intercepting Router Advertisements that the router thinks it is sending on the Thread interface and turning them into Thread-specific messaging. I am not opposed to this method, but I call it a "hack" because a more appropriate method would be to update OpenThread and/or DHCPv6 PD, etc. to interface more directly than via Router Advertisement messages that are never actually transmitted to any neighboring devices.

**Comment by gjc13** [comment]
> > > I think what is meant here is that we are using Router Advertisements to learn what prefixes are being assigned to the Thread network and translating them into Thread-specific messages. Is that correct?
> > 
> > 
> > This PR makes it possible to setup Thread network prefixes by forwarding certain `Router Advertisements` from Backbone to Thread network?
> 
> @simonlingoogle, couple notes:
> 
> 1. I don't think we are "forwarding" anything.
> 2. I think we are only processing Router Advertisements that are originated by the Thread border router and sent to the Thread interface. We are not processing Router Advertisements originated by any other device or received on any other interface.
> 
> What we are doing here is really more of a "hack". We are intercepting Router Advertisements that the router thinks it is sending on the Thread interface and turning them into Thread-specific messaging. I am not opposed to this method, but I call it a "hack" because a more appropriate method would be to update OpenThread and/or DHCPv6 PD, etc. to interface more directly than via Router Advertisement messages that are never actually transmitted to any neighboring devices.

The tricky thing on RA in Thread is that it sends to `ff02::1`. For use case in wifi and ethernet that's fine. Everyone is link-local. However for a mesh network like thread, link local devices is much more restricted so I have to handle the ra packets here.

**Comment by jwhui** [comment]
> > What we are doing here is really more of a "hack". We are intercepting Router Advertisements that the router thinks it is sending on the Thread interface and turning them into Thread-specific messaging. I am not opposed to this method, but I call it a "hack" because a more appropriate method would be to update OpenThread and/or DHCPv6 PD, etc. to interface more directly than via Router Advertisement messages that are never actually transmitted to any neighboring devices.
> 
> The tricky thing on RA in Thread is that it sends to `ff02::1`. For use case in wifi and ethernet that's fine. Everyone is link-local. However for a mesh network like thread, link local devices is much more restricted so I have to handle the ra packets here.

@gjc13 , I agree that [RFC 4861](https://tools.ietf.org/html/rfc4861) Router Advertisements are not applicable to Thread networks. However, my comment was more about the title and description of this PR that describes the proposed method as "forwarding" RAs, which I do not believe it is.

Also, I wonder if there is a better way to obtain the prefix information than from intercepting RAs sent to the Thread network. `wpantund` learns prefixes assigned to the Thread network by inspecting the addresses assigned to the interface. This is what https://github.com/openthread/ot-br-posix/pull/113 dependent on. However, I understand that simply observing prefix of addresses assigned to the network interface does not provide all the information. I accept that intercepting the RAs is one way to retrieve all the information we desire.

**Comment by abtink** [comment]
I also want to echo @jwhui comment. I think simply using APIs to add prefix/routes is more straight-forward and clear (from perspective of OT posix code). 

I think the next layers (e.g., `otbr` or maybe a new module) should be responsible for managing address/prefixes/routes which can be use-case/app dependent (e.g., such module can then listen and interprent RA's on other network interface and act accordingly (add/remove route/prfix on OT side)).

----

> In wpantund, we will add a prefix for any external addresses on the interface and ask users to statically config whether to treat the prefix as SLAAC or DHCP. This change can be viewed as an enhancement to the previous approach.

Regarding above, `wpantund` provides APIs to allow user to add any prefix/route (with any set of flags) they want. That's the main way we intend users to update prefix/route info in network data (on NCP).

The feature that auto-adds a prefix based on an IPv6 address added by user on `wpan0` interface is not intended as the way for users to add prefix. It's rather a helper feature to make it easier for users. 
- If user adds an address and expects it to be usable within the Thread network (i.e. other Thread node can talk to the device using this address) we would need the address prefix to be added as well (so other node know this is a on-mesh prefix and do address query for it). This feature just makes it easier for this case.
- The whole feature is optional and available when addresses between `wpan0` interface are auto added to NCP (`"Daemon:IPv6:AutoUpdateInterfaceAddrsOnNCP"` is enabled)
- The options to mark auto-added prefix as SLAAC or not, were add-ons intended for for an internal project/usecase and for debugging/testing only during earlier product development and not as a way to add an SLAAC prefix within Thread network.

**Comment by bukepo** [comment]
I agree it will be straight-forward for some users to directly add/remove prefixes or addresses. But if we only support that way, the user experience will be different from Wi-Fi/Ethernet. OpenThread daemon has been trying to give users a **native network interface experience similar to Wi-Fi and Ethernet**. If RA can be handled by those nodes automatically, then I tend make Thread nodes behave similar, so I think it makes sense to let Thread border routers convert RAs to the language Thread network understands.

Besides, we already made adding/removing addresses in netif in this project, I think the route and prefix should be in the same layer, which **only talks to the kernel** with no other dependencies. I intended to let ot-br-posix host stuffs for specific Linux distributions, such as OpenWrt, raspberry PI, beaglebone and some internal platforms, so that OpenThread itself will not be bothered by too much user space dependencies. Thoughts? @jwhui @abtink

**Comment by abtink** [comment]
> I agree it will be straight-forward for some users to directly add/remove prefixes or addresses. But if we only support that way, the user experience will be different from Wi-Fi/Ethernet. OpenThread daemon has been trying to give users a native network interface experience similar to Wi-Fi and Ethernet. If RA can be handled by those nodes automatically, then I tend make Thread nodes behave similar, so I think it makes sense to let Thread border routers convert RAs to the language Thread network understands.

I see a key difference here. On Wi-Fi/Ethernet, some entity on a router (e.g., the RA daemon) generates the RA msgs which are then broadcast on WiFi interface so that all the other WiFi devices within the same subnet receive and parse the RA and apply/adopt the router/prefix info from it. This is not applicable to Thread.

Some questions which may help clarify the use (help me understand): 

- The code in this PR is adding a listener on "wpan" interface itself
- What entity is creating the RA message and emitting them on "wpan" interface?
   - Most likely this entity is on Thread (border) router itself. In that case,  would it not be easier to directly use APIs to add prefix/routes (compared to it generating and sending an RA msg to wpan interface which is then snooped and converted to calling APIs to add prefix/routes on Thread network).
   - In case of remote entity (not on same device), do we want to allow RA messages over Thread network? I would guess no. Also I don't think it would correct to forward RA from one subnet to another (we cannot forward RA from WiFi subnet to "wpan" as they are different subnets with different prefixes). So how is intended to be used in this case?

**Comment by gjc13** [comment]
> > I agree it will be straight-forward for some users to directly add/remove prefixes or addresses. But if we only support that way, the user experience will be different from Wi-Fi/Ethernet. OpenThread daemon has been trying to give users a native network interface experience similar to Wi-Fi and Ethernet. If RA can be handled by those nodes automatically, then I tend make Thread nodes behave similar, so I think it makes sense to let Thread border routers convert RAs to the language Thread network understands.
> 
> I see a key difference here. On Wi-Fi/Ethernet, some entity on a router (e.g., the RA daemon) generates the RA msgs which are then broadcast on WiFi interface so that all the other WiFi devices within the same subnet receive and parse the RA and apply/adopt the router/prefix info from it. This is not applicable to Thread.
> 
> Some questions which may help clarify the use (help me understand):
> 
> * The code in this PR is adding a listener on "wpan" interface itself
> * What entity is creating the RA message and emitting them on "wpan" interface?
>   
>   * Most likely this entity is on Thread (border) router itself. In that case,  would it not be easier to directly use APIs to add prefix/routes (compared to it generating and sending an RA msg to wpan interface which is then snooped and converted to calling APIs to add prefix/routes on Thread network).
>   * In case of remote entity (not on same device), do we want to allow RA messages over Thread network? I would guess no. Also I don't think it would correct to forward RA from one subnet to another (we cannot forward RA from WiFi subnet to "wpan" as they are different subnets with different prefixes). So how is intended to be used in this case?

An alternative is to add a filter for all incoming packets and update apply prefixes sent from `radvd` and `dhcpd`. The pro of such approach is that it allows users to reuse these common network management daemons rather than applying Thread specific configs. Is this the preferred solution?

**Comment by gjc13** [comment]
Replaced by 1.3 efforts.

**Pull request closed without merging**